### DISCUSSION TITLE
About connecting to a DB server developed based on PostgreSQL.

### DISCUSSION THREAD

**Discussion by me-aikawa** [sqlalchemy/sqlalchemy#9378] [Usage Questions] [answered]
Hi, 

I would like to connect to a DB server developed based on PostgreSQL ([PGSpider](https://github.com/pgspider/pgspider)) with SQLAlchemy v1.4.27. 

I have executed the following code and got an error.

```python
from sqlalchemy import create_engine, inspect
engine = create_engine("postgresql://pgspider:xxxxx@/pgspiderdb?host=localhost&port=4813")

with engine.connect() as conn:
     conn.execute(text("CREATE TABLE some_table (x int, y int)"))
     conn.execute(text("INSERT INTO some_table (x, y) VALUES (:x, :y)"),[{"x": 1, "y": 1}, {"x": 2, "y": 4}])
```

```sh
AssertionError: Could not determine version from string 'PGSpider 15beta2 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44), 64-bit'
```

I found out that it is caused by the following checks made within the _get_server_version_info method of the PGDialect class.

https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/postgresql/base.py#L3213
```python
def _get_server_version_info(self, connection):
        v = connection.exec_driver_sql("select pg_catalog.version()").scalar()
        m = re.match(
            r".*(?:PostgreSQL|EnterpriseDB) "
            r"(\d+)\.?(\d+)?(?:\.(\d+))?(?:\.\d+)?(?:devel|beta)?",
            v,
        )
```

Then, I changed `PG_VERSION_STR` in PGSpider's source code from `PGSpider 15beta2 ... ` to `PostgreSQL 15beta2 ...` and builded PGSpider. Now the error no longer occurs. However, I would like to use SQLAlchemy to connect to PGSpider without modifying `PG_VERSION_STR` if possible.

If there is a way to resolve this situation, could you please let me know?
For example, is there an option to skip `_get_server_version_info()`?

**Comment by CaselIT** [comment, marked as answer]
Hi,

The best option is to create a new dialed for that particular incantation of postgresql. That check is there since there are a lot of more or less compatible postgresql implementation, and supporting them all with a single implementation is unfeasible.

You can create a new dialect that can subclass the internal one to add any required change to support that database

**Reply by me-aikawa** [reply]
Thank you for responding so quickly, @CaselIT
Is there any documentation or sample code on the development and use of a new dialects?

**Reply by CaselIT** [reply]
Hi, general documentation is in https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst

You can look at the 3rd party dialect list for examples https://docs.sqlalchemy.org/en/20/dialects/index.html#external-dialects there are some dialects for pg like db listed too

**Reply by me-aikawa** [reply]
@CaselIT 
Thank you for the information. I'll try it.

**Reply by CaselIT** [reply]
Feel free to ask questions if you need support in something on the dialect side.
